src/vos/vos_obj_flat.h

```diff
@@ -0,0 +1,128 @@
+/**
+ * (C) Copyright 2023-2023 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * Object flatten related API and structures
+ * src/vos/vos_obj_flat.h
+ */
+
+#ifndef __VOS_OBJ_FLAT_H__
+#define __VOS_OBJ_FLAT_H__
+
+#include "vos_layout.h"
+
+enum {
+	VOF_DKEY	= 0x1,
+	VOF_AKEY	= 0x2,
+	VOF_SINGV	= 0x3,
+	VOF_EXT		= 0x4,
+};
+
+#define VOF_KEY_INLINE_SZ	(4)
+#define VOF_SINGV_INLINE_SZ	(8)
+
+/** dkey/akey/value for VOS flattened object, 20B each */
+struct vof_item_df {
+	/** type (dkey/akey/sing-value/array-ext) */
+	uint16_t				ofk_type;
+	/** number of checksums */
+	uint16_t				ofk_cs_nr;
+	/** size of the item, for single-value is the local record size */
+	uint32_t				ofk_size;
+	union {
+		/* dkey/akey */
+		struct {
+			union {
+				/** offset of the key */
+				uint32_t	ofk_key_off;
+				/** inline key (for short key and no key csum) */
+				uint8_t		ofk_key[VOF_KEY_INLINE_SZ];
+			};
+			/** offset of children (akey or value) */
+			uint32_t		ofk_child_off;
+			/** number of children */
+			uint16_t		ofk_child_nr;
+			/** type of children */
+			uint16_t		ofk_child_type;
+		};
+		/* array extent */
+		struct {
+			/** index of the value extent */
+			uint64_t		ofk_ext_idx;
+			/** offset of value */
+			uint32_t		ofk_ext_off;
+		};
+		/* single-value */
+		struct {
+			/* global record size of the single-value */
+			uint32_t		ofk_singv_gsize;
+			union {
+				/** offset of the single-value */
+				uint32_t		ofk_singv_off;
+				/** inline data (for short single-value and no csum) */
+				uint8_t			ofk_singv[VOF_SINGV_INLINE_SZ];
+			};
+		};
+	};
+} __attribute__ ((aligned(4)));
+
+/** flattened object durable format */
+struct vos_obj_flat_df {
+	/** format version */
+	uint16_t	ofd_version;
+	uint16_t	ofd_pad;
+	/** number of bytes per index(record) */
+	uint32_t	ofd_inob;
+	/** number of bytes used to generate each csum */
+	uint32_t	ofd_csum_chunk_sz;
+	/** type of the csum */
+	uint16_t	ofd_csum_type;
+	/** length of each csum in bytes */
+	uint16_t	ofd_csum_len;
+	/** flatten epoch */
+	daos_epoch_t	ofd_flat_epoch;
+	/** number of dkey */
+	uint16_t	ofd_dkey_nr;
+	/** number of vof_item_df in payload */
+	uint16_t	ofd_item_nr;
+	/** number of bytes of payload */
+	uint32_t	ofd_len;
+	/**
+	 * payload's layout -
+	 * struct vof_item_df[ofd_item_nr]
+	 * followed by each key and value's content and checksums.
+	 */
```

Comment:
do we want to have checksum for each individual item? I think one checksum per flattend object is fine, we can verify it while loading. It can save more space and make the code simpler?

Reply:
now in EV tree it stores csums for each entry (struct evt_entry::en_csum), and for single-value/dkey/akey all with its csums stored. When fetch, those csums can be read back and compare/transfer back to client to compare.
If for flattened obj we changed it to only store a whole csums for the obj_flat_df and verify it while loading, it means -
1) we need to calculate related csums when fetch the data every time to transfer back to client, 
2) seems can broke the end-to-end checksum assumption as the data-corruption between obj loading and csum re-calculate cannot be detected?

Reply:
I presume you would just need to minic the behavior.  Like if you had one checksum for the whole flat object but then recalculated the checksum for each checksum chunk you return to client